File: backend/services/order-service/internal/model/order.go
package model

import (
	"encoding/json"
	"time"
)

// Order 对应 orders 表，订单主表。
//
// 这里没有 order_type：一次支付可以只买咖啡、只买会员，也可以两样一起买，
// 「这一单是什么类型」由它有哪些 order_lines 决定（见迁移 001 的说明）。
type Order struct {
	ID       string  `db:"id"`
	OrderNo  string  `db:"order_no"`
	LegacyID *string `db:"legacy_id"`
	// 下单用户。nil = 这一单**没有用户**，只有线下刷卡机这一条路会这样（source=device）：
	// 钱在机器上收过了，我们只是把既成事实记下来，它不属于任何用户。
	//
	// 指针而不是空串：空串与「某个真的用户」长得一样，uuid 列上也根本存不进去——空串要靠
	// NULLIF 才会变成 NULL，任何一处忘了写就是 22P02。而「没有用户」在这条路上是**常态**，
	// 不是一个要绕开的边界。
	UserID *string `db:"user_id"`
	// 下单来源：miniapp=小程序直接下单，screen_qr=咖啡机屏幕选品后扫码下单，
	// device=线下刷卡机设备回调（partner-service 验签后经 gRPC 建单），
	// renewal=会员续费代扣（membership-service 在扣款成功之后经 gRPC 建单）。
	Source string `db:"source"`
	// 主状态机：pending_payment → paid → completed，或 pending_payment → cancelled / expired。
	Status string `db:"status"`
	// 履约汇总，事实由 fulfillment-service 的出杯任务产生；本库只接收事件更新这一个字段。
	// none 表示本单没有任何需要履约的行（纯会员订单）。
	FulfillmentStatus string `db:"fulfillment_status"`
	// 点位与咖啡机的值引用 + 下单当时的快照名/编号。点位名与设备编号是快照，
	// 改名换号以后历史订单仍能还原成用户当时看到的样子。
	StoreID   *string `db:"store_id"`
	StoreName string  `db:"store_name"`
	DeviceID  *string `db:"device_id"`
	DeviceNo  string  `db:"device_no"`
	// 屏幕选品的会话，只作值引用；选品快照落在饮品行上。
	SceneToken string `db:"scene_token"`
	// 金额，单位分。payable = original - discount 是定义，由 orders_payable_matches 约束保证。
	OriginalAmount int64 `db:"original_amount"`
	DiscountAmount int64 `db:"discount_amount"`
	PayableAmount  int64 `db:"payable_amount"`
	PaidAmount     int64 `db:"paid_amount"`
	RefundedAmount int64 `db:"refunded_amount"`
	// 下单时的会员资格快照。结构归 membership-service，本库只存一份副本用于还原当时的会员价。
	MembershipID       *string         `db:"membership_id"`
	MembershipSnapshot json.RawMessage `db:"membership_snapshot"`
	// 这一单承诺发多少张福卡（基础赠送 + 各加购活动加赠）。只存承诺，不存发放结果。
	FortuneCardsExpected int             `db:"fortune_cards_expected"`
	FortuneCardSnapshot  json.RawMessage `db:"fortune_card_snapshot"`
	// 主支付渠道与支付单号（payment-service 的值引用）。逐笔出资分摊在 order_payment_lines。
	PaymentMethod      string     `db:"payment_method"`
	PaymentNo          string     `db:"payment_no"`
	PaidAt             *time.Time `db:"paid_at"`
	FinishedAt         *time.Time `db:"finished_at"`
	CancelledAt        *time.Time `db:"cancelled_at"`
	CancellationReason string     `db:"cancellation_reason"`
	// 待支付超时时间，关单扫描按它扫。
	ExpiresAt *time.Time `db:"expires_at"`
	Remark    string     `db:"remark"`
	// 对方单号，设备回调的幂等键（orders_third_party_order_no_key）。空串表示这一单没有
	// 对方单号——小程序与屏幕扫码那两条路都是空的。
	ThirdPartyOrderNo string `db:"third_party_order_no"`
	// 下单请求号：同一个 request_id 只能落一个订单（orders_request_id_key）。
	RequestID string    `db:"request_id"`
	CreatedAt time.Time `db:"created_at"`
	UpdatedAt time.Time `db:"updated_at"`
}

// OrderStatus 是 orders.status 的取值全集，与迁移里的 CHECK 逐字一致。
const (
	OrderStatusPendingPayment = "pending_payment"
	OrderStatusPaid           = "paid"
	OrderStatusCompleted      = "completed"
	OrderStatusCancelled      = "cancelled"
	OrderStatusExpired        = "expired"
	OrderStatusRefunding      = "refunding"
	OrderStatusRefunded       = "refunded"
)

// FulfillmentStatus 是 orders.fulfillment_status 的取值全集。
const (
	FulfillmentPending   = "pending"
	FulfillmentMaking    = "making"
	FulfillmentReady     = "ready"
	FulfillmentCompleted = "completed"
	FulfillmentFailed    = "failed"
	FulfillmentCancelled = "cancelled"
	// FulfillmentNone：本单没有任何需要履约的行（纯会员订单）。没有这个值，
	// 会员订单会永远停在 pending，后台「待制作」列表里就会混进一堆做不完的会员单。
	FulfillmentNone = "none"
)

// OrderSource 是 orders.source 的取值全集，与迁移里的 CHECK 逐字一致。
const (
	SourceMiniapp  = "miniapp"
	SourceScreenQR = "screen_qr"
	// SourceDevice 是线下刷卡机：partner-service 验完厂商签名后经 gRPC 建单，没有用户、
	// 没有支付单，**直接落成已支付**（钱在机器上已经收过了，见 order/005）。
	SourceDevice = "device"
	// SourceRenewal 是**会员续费**：连续包月每期代扣成功之后，由 membership-service 经 gRPC
	// 建单，同样直接落成已支付（钱是微信代扣收的，我们手里只有渠道流水号）。
	//
	// 这一条**是这三条「钱已在别处收过」的路里唯一有用户的一条**（user_id 非空）。它买的是
	// 会员时长，没有任何需要履约的行——fulfillment_status 恒为 none（纯会员订单）。
	//
	// 为什么不复用 miniapp：那条路的语义是「用户在小程序里点了下单」，而续费根本没有这一下
	// ——没有人点，是到期自动扣的。后台列表的来源筛选、以及按来源看的每一张报表，靠的都是
	// 这一列，把它们混在一起就再也分不开了（见 order/009）。
	//
	// **订单号不另起前缀**（老系统续费单是 SUB 开头）：单号格式是收单渠道的约定
	// （3CYM + 时间 + 随机，见 generateOrderNo），而「这是续费」这件事已经由本列说清楚了
	// ——用前缀表达的代价是所有认单号格式的地方都要多认一种。
	SourceRenewal = "renewal"
)

// IsOrderSource 判断一个值是不是合法的下单来源，与 orders_source_check 逐字一致。
//
// 后台列表的 source 筛选用它**在进 SQL 之前**挡下打错的值（照 IsPaymentStatus 那条）：
// 一个不在词表里的值送进 SQL 不报错，只会安静地返回空列表，而运营会把它当成「这个来源
// 真的没有订单」。
func IsOrderSource(source string) bool {
	switch source {
	case SourceMiniapp, SourceScreenQR, SourceDevice, SourceRenewal:
		return true
	default:
		return false
	}
}

File: merchant-web/src/services/order.ts
import { request } from '@umijs/max';
import type { PageQuery, PageResult } from './pagination';

/**
 * 商户端订单接口。字段名与 order-service dto（internal/dto/order.go、after_sale.go）的
 * json tag 逐字对应，同时也是页面上的 dataIndex——改一处要同批改三处，否则 TypeScript 不
 * 报错，只是整列空白。
 *
 * 金额一律是「分」的整数（int64 到了 JS 里就是 number）。这一层不换算，页面按元展示。
 *
 * 这里没有「按门店筛」的入参：能看见哪些单由服务端按令牌实时解析出来的数据范围决定。
 * 售后**不在商户端范围内**——它不是独立入口，只作为订单详情的一部分随订单一起受范围约束。
 */

// ——— 枚举 ———
// 取值来自 order-service 的 model 常量（各表的 CHECK 约束），不是接口给的——接口回的
// 就是库里那些英文码。所以**改枚举必须同时改迁移和这里**，加了新码而这里没登记，界面上
// 就退回显示原始码。文案表在 services/orderLabels.ts。

export type OrderStatus =
  | 'pending_payment'
  | 'paid'
  | 'completed'
  | 'cancelled'
  | 'expired'
  | 'refunding'
  | 'refunded';

export type FulfillmentStatus =
  | 'none'
  | 'pending'
  | 'making'
  | 'ready'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type OrderSource = 'miniapp' | 'screen_qr';
export type OrderLineType = 'drink' | 'addon' | 'membership';
export type PaymentLineStatus = 'reserved' | 'succeeded' | 'failed' | 'released' | 'reversed';

export type ActorType = 'user' | 'merchant' | 'admin' | 'system';
export type AggregateType = 'order' | 'order_line' | 'after_sale' | 'payment_line';

export type AfterSaleStatus =
  | 'pending'
  | 'approved'
  | 'rejected'
  | 'refunding'
  | 'refunded'
  | 'failed'
  | 'cancelled';

export type AfterSaleScope = 'all' | 'drink' | 'addon' | 'membership';

// ——— 订单 ———

/** 列表里的一行。不带行明细与出资分摊——那是详情的事。 */
export type OrderSummary = {
  id: string;
  orderNo: string;
  userId: string;
  source: OrderSource;
  status: OrderStatus;
  fulfillmentStatus: FulfillmentStatus;
  storeId: string | null;
  storeName: string;
  deviceId: string | null;
  deviceNo: string;
  originalAmount: number;
  discountAmount: number;
  payableAmount: number;
  paidAmount: number;
  refundedAmount: number;
  /**
   * 支付方式的 code（payment-service 目录里的，如 ums_h5_alipay / coffee_bean）。
   *
   * 类型是 string 而不是联合类型：这张值域表由支付服务持有、会随后端加支付方式而长，把六条
   * code 抄成联合类型只会让「后端的第七种」在编译期变成一个假错误。设备单还会在这里写
   * card_pay / pickup_code（订单侧自造的标签，没有支付单）。
   *
   * 界面上按已知的那几条翻成中文，认不出来的原样回显（见 orderLabels.paymentMethodLabel）。
   */
  paymentMethod: string;
  paymentNo: string;
  paidAt: string | null;
  finishedAt: string | null;
  cancelledAt: string | null;
  cancellationReason: string;
  expiresAt: string | null;
  remark: string;
  /** 只读派生：这一单有没有饮品行 / 加购行 / 会员行，列表「构成」列用。 */
  hasDrinkLine: boolean;
  hasAddonLine: boolean;
  hasMembershipLine: boolean;
  /** 下单时承诺赠送的福卡张数。 >0 时审核退款要有人确认没抽过奖（商户端只读，只展示）。 */
  fortuneCardsExpected: number;
  /**
   * 取杯号：这一单饮品行那个短号。没付成功或没有饮品行时为 null。
   *
   * 用户侧叫取杯号，取杯口屏幕上叫取杯码，是**同一个值**，不是凭据（见 migrations/order）。
   */
  pickupCode: string | null;
  createdAt: string;
  updatedAt: string;
};

/** 订单行。 */
export type OrderLine = {
  id: string;
  lineNo: number;
  lineType: OrderLineType;
  itemId: string | null;
  itemCode: string;
  itemName: string;
  itemImage: string;
  quantity: number;
  originalUnitPrice: number;
  unitPrice: number;
  priceDiscountAmount: number;
  discountAmount: number;
  payableAmount: number;
  couponId: string | null;
  couponDiscountAmount: number;
  specs: unknown;
  selectionSnapshot: unknown;
  campaignId: string | null;
  campaignSnapshot: unknown;
  membershipPlanSnapshot: unknown;
  deviceId: string | null;
  deviceOrderNo: string;
  fulfillmentTaskNo: string;
  /** 取杯号，只有饮品行有，支付成功时生成。 */
  pickupCode: string | null;
  remark: string;
  createdAt: string;
  updatedAt: string;
};

/**
 * 一次出资分摊。
 *
 * 没有渠道流水号：`provider_transaction_id` 后端有意不映射（对账凭据，要看去 payment-service
 * 查），所以这里既没有这个字段，页面也不该去找。
 */
export type OrderPaymentLine = {
  id: string;
  lineNo: number;
  /** 支付方式的 code，与 orders.payment_method 同一个值（词表已退场，见 migrations/order）。 */
  lineType: string;
  amount: number;
  status: PaymentLineStatus;
  paymentNo: string;
  failureCode: string;
  accountEntryId: string | null;
  succeededAt: string | null;
  reversedAt: string | null;
  createdAt: string;
};

/** 一条状态流水。前两个字段一起回答「谁的状态从什么变成了什么」。 */
export type OrderTransition = {
  id: string;
  aggregateType: AggregateType;
  aggregateId: string;
  fromStatus: string;
  toStatus: string;
  reason: string;
  actorType: ActorType;
  actorId: string | null;
  createdAt: string;
};

/** 按行退时被退的那一行（整单退为 null）。 */
export type AfterSaleLineRef = {
  id: string;
  lineNo: number;
  lineType: OrderLineType;
  itemName: string;
  quantity: number;
  payableAmount: number;
};

/**
 * 售后单。后三个字段是只读派生（那三个数不在售后单上，是从订单和订单行上取的），
 * 审核福卡规则要看的就是它们。
 */
export type AfterSale = {
  id: string;
  afterSaleNo: string;
  orderId: string;
  orderNo: string;
  userId: string;
  /** 目前只有 refund。 */
  type: string;
  scope: AfterSaleScope;
  orderLineId: string | null;
  status: AfterSaleStatus;
  reason: string;
  /** 用户上传的凭证图片 URL 列表，可能是 null。 */
  images: string[] | null;
  refundAmount: number;
  refundNo: string;
  failureCode: string;
  reviewedBy: string | null;
  reviewedAt: string | null;
  reviewRemark: string;
  refundedAt: string | null;
  createdAt: string;
  updatedAt: string;
  fortuneCardsExpected: number;
  fortuneCardSnapshot: unknown;
  orderLine: AfterSaleLineRef | null;
};

/** 订单详情 = 列表那一行 + 明细 + 出资 + 流水 + 这一单的售后记录。 */
export type OrderDetail = OrderSummary & {
  sceneToken: string;
  membershipId: string | null;
  membershipSnapshot: unknown;
  fortuneCardSnapshot: unknown;
  lines: OrderLine[];
  paymentLines: OrderPaymentLine[];
  transitions: OrderTransition[];
  afterSales: AfterSale[];
};

// ——— 接口 ———

/**
 * 订单筛选。只认 status / orderNo / source / createdFrom / createdTo——后端
 * MerchantOrderQuery 里**没有 keyword**，orderNo 是精确匹配（`=`），所以调用前要 trim，
 * 也不要指望它做模糊。
 */
export type MerchantOrderQuery = PageQuery & {
  orderNo?: string;
  status?: OrderStatus;
  source?: OrderSource;
  /** 创建时间的闭开区间 [createdFrom, createdTo)，必须是带时区的 RFC3339（用 toRFC3339 转）。 */
  createdFrom?: string;
  createdTo?: string;
};

export async function listOrders(params?: MerchantOrderQuery) {
  return request<PageResult<OrderSummary>>('/api/v1/merchant/orders', { params });
}

/** 订单详情。范围外与不存在都回 404，服务端不区分这两件事。 */
export async function getOrder(id: string) {
  return request<OrderDetail>(`/api/v1/merchant/orders/${id}`);
}
